### PR TITLE
Limit pod_name length to HOST_NAME_MAX

### LINKED ISSUE #36326: Task hostname does not reflect pod name when it's too long
### Apache Airflow version

2.8.0

### If "Other Airflow 2 version" selected, which one?

_No response_

### What happened?

When a task has a long name (or it's inside one or many task group(s) with long names), the name of the pod that is created for the task is different than the task hostname (from Airflow Metadata DB)

### What you think should happen instead?

The task hostname should be the same as the pod name

### How to reproduce

1. Run the following DAG
  ```python
  import time
from airflow.decorators import task, task_group
from airflow.models import DAG
from datetime import datetime, timedelta

with DAG(
    dag_id="test_long_task_name",
    schedule=None,
    default_args={
        "depends_on_past": False,
        "wait_for_downstream": False,
        "start_date": datetime(2023, 12, 20),
        "retry_delay": timedelta(seconds=10),
        "retries": 1,
    },
    catchup=False,
    max_active_runs=1,
    render_template_as_native_obj=True,
) as dag:

    @task_group
    def task_group_with_long_name():
        @task_group
        def inner_task_group_with_long_name():
            @task
            def task1():
                pass

            @task
            def task2():
                pass

            task1()
            task2()

        inner_task_group_with_long_name()

    task_group_with_long_name()
  ```
image

2. Check the hostname property for both tasks:
image
image

3. Check the pod name for both tasks (scheduler logs):
```
[2023-12-20T09:45:05.497+0000] {kubernetes_executor_utils.py:396} INFO - Creating kubernetes pod for job is TaskInstanceKey(dag_id='test_long_task_name', task_id='task_group_with_long_name.inner_task_group_with_long_name.task2', run_id='manual__2023-12-20T09:45:00+00:00', try_number=1, map_index=-1), with pod name test-long-task-name-task-group-with-long-name-inner-task-group-with-lon-efj7zr6z, annotations: <omitted>
```

```
[2023-12-20T09:45:05.776+0000] {kubernetes_executor_utils.py:396} INFO - Creating kubernetes pod for job is TaskInstanceKey(dag_id='test_long_task_name', task_id='task_group_with_long_name.inner_task_group_with_long_name.task1', run_id='manual__2023-12-20T09:45:00+00:00', try_number=1, map_index=-1), with pod name test-long-task-name-task-group-with-long-name-inner-task-group-with-lon-diwm37yp, annotations: <omitted>
```

## Task1
- hostname: `test-long-task-name-task-group-with-long-name-inner-task-group`
- pod name: `test-long-task-name-task-group-with-long-name-inner-task-group-with-lon-diwm37yp`
## Task2
- hostname: `test-long-task-name-task-group-with-long-name-inner-task-group`
- pod name: `test-long-task-name-task-group-with-long-name-inner-task-group-with-lon-efj7zr6z`

### Operating System

Airflow in Kubernetes (`apache/airflow:2.8.0-python3.11`)

### Versions of Apache Airflow Providers

apache-airflow-providers-cncf-kubernetes==7.11.0

### Deployment

Official Apache Airflow Helm Chart

### Deployment details

_No response_

### Anything else?

This issue also happens in previous Airflow versions.

After some research, I've found out the following:
- The function that is used to determine the hostname is defined with the [hostname_callable](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#hostname-callable) parameter.
- There's a maximum size for this hostname [[source](https://stackoverflow.com/questions/8724954/what-is-the-maximum-number-of-characters-for-a-host-name-in-unix)]. In the Airflow image, this limit is set to **64**.
- In the k8s provider, the pod name maximum size is set to 80 [[source](https://github.com/apache/airflow/blob/2.8.0/airflow/providers/cncf/kubernetes/kubernetes_helper_functions.py#L62)]

So, for me, there are two possible approaches to solve this issue
1. Modify the k8s provider pod name maximum size to 63, matching `HOST_NAME_MAX` minus 1.
2. Increase the `HOST_NAME_MAX` limit in the Airflow image.

I believe the first one is the safest

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by csp33**
Closes #36326

### CODE REVIEW

**Comment by potiuk** [comment]
How do you plan to tackle backwards compatibilty. Imagine you are upgrading Airflow and thera are long running Pods still running being adopted by the new Airflow with limited length/ What is your plan to do with it?

**Comment by dirrao** [comment]
> How do you plan to tackle backwards compatibilty. Imagine you are upgrading Airflow and thera are long running Pods still running being adopted by the new Airflow with limited length/ What is your plan to do with it?

Pod adoption is seamless. As we are not using pod name to deduce anything. Its just a reference to identify the dag and task from pod while debugging.

**Comment by dirrao** [comment]
@csp33 
Nice work. we have seen problem where K8 cluster not able to resolve/update DNS because of when they exceeds max length 63.

**Comment by csp33** [comment]
> How do you plan to tackle backwards compatibilty. Imagine you are upgrading Airflow and thera are long running Pods still running being adopted by the new Airflow with limited length/ What is your plan to do with it?

K8s executor does not use the pod name to adopt previous TIs, it uses the `queued_by_job_id` field instead:
https://github.com/apache/airflow/blob/73cd0d5f8e99f93d2ff90a7cca04e6cdf086c359/airflow/providers/cncf/kubernetes/executors/kubernetes_executor.py#L581

**Review by dirrao — APPROVED** [review]
Awesome work. LGTM.

**Comment by potiuk** [comment]
> > How do you plan to tackle backwards compatibilty. Imagine you are upgrading Airflow and thera are long running Pods still running being adopted by the new Airflow with limited length/ What is your plan to do with it?
> 
> K8s executor does not use the pod name to adopt previous TIs, it uses the `queued_by_job_id` field instead:
> 
> https://github.com/apache/airflow/blob/73cd0d5f8e99f93d2ff90a7cca04e6cdf086c359/airflow/providers/cncf/kubernetes/executors/kubernetes_executor.py#L581

Cool

File: airflow/providers/cncf/kubernetes/kubernetes_helper_functions.py [hunk 1]
```diff
@@ -59,15 +59,13 @@ def create_pod_id(
     dag_id: str | None = None,
     task_id: str | None = None,
     *,
-    max_length: int = 80,
```
Comment (potiuk) [inline airflow/providers/cncf/kubernetes/kubernetes_helper_functions.py:62]: Can you please extract this as constant while you are at it?

File: airflow/kubernetes/pre_7_4_0_compatibility/pod_generator.py [hunk 2]
```diff
@@ -72,7 +72,7 @@ def rand_str(num):
     return "".join(secrets.choice(alphanum_lower) for _ in range(num))
 
 
-def add_pod_suffix(pod_name: str, rand_len: int = 8, max_len: int = 80) -> str:
+def add_pod_suffix(pod_name: str, rand_len: int = 8, max_len: int = 63) -> str:
```
Comment (potiuk) [inline airflow/kubernetes/pre_7_4_0_compatibility/pod_generator.py]: Do we want to change it here? This just for pre-7.4.0 compatibility it seems?

Comment (csp33) [reply airflow/kubernetes/pre_7_4_0_compatibility/pod_generator.py]: Okay, I'll keep that file unchanged ;)

File: airflow/providers/cncf/kubernetes/kubernetes_helper_functions.py (hunk 1 above)
Comment (csp33) [reply airflow/providers/cncf/kubernetes/kubernetes_helper_functions.py:62]: done!

**Review by potiuk — APPROVED** [review]

**Pull request merged**